### ISSUE TITLE
Sort out design time interaction with Startup in ASP.NET Core

### ISSUE THREAD

**Issue by rowanmiller** [aspnet/EntityFramework#4546]
Our tooling currently runs code in Startup.cs to get the context from services. This is messy though because our design time environment is not the same as the environment the app really runs in. We need to revisit this.

As part of this we need to try and remove ASP.NET specific dependencies from our general commands package so you don't get references to `Microsoft.AspNetCore.Hosting.*` etc in a UWP/WinForms/WPF project (see https://github.com/aspnet/EntityFramework/issues/4830).

Related issues:
 * https://github.com/aspnet/EntityFramework/issues/4494
 * https://github.com/aspnet/EntityFramework/issues/4367

**Comment by natemcmaster** [comment]
See also https://github.com/aspnet/EntityFramework/pull/5037. In this change, we specifically workaround limitations in UWP and startupinvoker. We will need a different abstraction when/if we stop cross-compiling to UWP. #5045

**Comment by natemcmaster** [comment]
We will likely need to react to this: https://github.com/aspnet/Hosting/pull/714
cc @divega

**Comment by natemcmaster** [comment]
Related issue: https://github.com/aspnet/EntityFramework/issues/5304#issuecomment-217956745

Recent Hosting templates have changed to loading configuration in `Program.Main`. EF can't (and shouldn't) call `Program.Main` at design-time, yet we still need to load configuration to get connection strings.

**Comment by bricelam** [comment]
We discussed this today. We came up with a solution that would keep some of the ASP.NET-specific code outside of the main tooling package, but decided that the cost and complexity of implementing it wasn't worth the marginal value it would add.

As part of #5334, we are going to author and structure the packages in a way that reduces the amount of dependencies that get pulled into your project and utilize NuGet & CLI features to prevent the ones we do pull in from getting deployed with your application.

I'm closing this issue and splitting off the remaining work into #5616, #5615 & #5617
